includes/Hooks.php

```diff
@@ -148,6 +152,90 @@ public static function factory(
 		);
 	}
 
+	/**
+	 * Add nofollow to redlinks to prevent search engines from indexing these links to save crawl budget because redlinks are noindex 404 pages
+	 * 
+	 * @see https://github.com/marohh/mediawikiRemoveRedlinks/blob/master/includes/RemoveRedlinks.php
+	 * 
+	 * @param LinkRenderer $linkRendere The LinkRenderer object
+	 * @param LinkTarget $target The target of the link
+	 * @param boolean $isKnown Whether the page exists or not
+	 * @param HtmlArmor|string $text The contents of the <a> tag
+	 * @param string[] &$attribs Link attributes
+	 * @param string &$ret The value to return if the hook returns false
+	 * 
+	 * @return bool
+	 */
+	public function onHtmlPageLinkRendererEnd(
+		$linkRenderer, 
+		$target,
+		$isKnown,
+		&$text,
+		&$attribs,
+		&$ret
+	) {
+		if ( $isKnown || $target->isExternal() ) {
+			return true;
+		}
+
+		$attribs['rel'] = 'nofollow';
+
+		return true;
+	}
+
+	/**
+	 * Add noindex to some pages for SEO purposes. Indexing pages that are not valuable to searchers is bad for SEO, so we'll reduce the indexing of such pages.
+	 * 
+	 * @see https://ahrefs.com/blog/content-pruning/
+	 * @see https://gitlab.com/hydrawiki/extensions/seo/-/blob/master/SEOHooks.php?ref_type=heads
+	 * 
+	 * @param OutputPage $out The OutputPage object
+	 * @param Skin $skin The Skin object that will be used to generate the page
+	 */
+	public function onBeforePageDisplay( $out, $skin ) : void {
+		$noIndexNamespaces = [
+			NS_SPECIAL,
+			NS_CATEGORY_TALK,
+			NS_MEDIAWIKI,
+			NS_MEDIAWIKI_TALK,
+			NS_USER,
+			NS_USER_TALK
+		];
+
+		if ( in_array( $out->getTitle()->getNamespace(), $noIndexNamespaces ) ) {
+			$out->setRobotPolicy( 'noindex,nofollow' );
+			return;
+		}
+
+		$blockedURLParamKeys = [
+			'curid', 'diff', 'from', 'group', 'mobileaction', 'oldid',
+			'printable', 'profile', 'redirect', 'redlink', 'stableid',
+			'veaction', 'action'
+		];
+
+		$blockedURLParamKeyValuePairs = [
+			'feed' => ['rss'],
+			'limit' => ['500'],
+			'title' => [
+				'Category:Noindexed_pages',
+				'Category:Noindexed pages',
+				'Category:Noindexed%20pages'
+			]
+		];
+
+		foreach ( $out->getRequest()->getValues() as $key => $value ) {
+			if ( in_array($key, $blockedURLParamKeys) ) {
+				$out->setRobotPolicy( 'noindex,nofollow' );
+				return;
+			}
+
+			if ( isset($blockedURLParamKeyValuePairs[$key] ) && in_array( $value, $blockedURLParamKeyValuePairs[$key] ) ) {
```

Comment:
Spacing

Reply:
Actually I think this whole check could be simplifed to `in_array( $value, $blockedURLParamKeyValuePairs[$key] ?? [] )`